Source/Core/UICommon/DiscordPresence.cpp

```diff
@@ -15,36 +20,174 @@
 
 namespace Discord
 {
+#ifdef USE_DISCORD_PRESENCE
+static Handler* event_handler = nullptr;
+static const char* username = "";
+
+static void HandleDiscordReady(const DiscordUser* user)
+{
+  username = user->username;
+}
+
+static void HandleDiscordJoinRequest(const DiscordUser* user)
+{
+  if (event_handler == nullptr)
+    return;
+
+  const std::string discord_tag = StringFromFormat("%s#%s", user->username, user->discriminator);
+  event_handler->DiscordJoinRequest(user->userId, discord_tag, user->avatar);
+}
+
+static void HandleDiscordJoin(const char* join_secret)
+{
+  if (event_handler == nullptr)
+    return;
+
+  if (Config::Get(Config::NETPLAY_NICKNAME) == Config::NETPLAY_NICKNAME.default_value)
+    Config::SetBaseOrCurrent(Config::NETPLAY_NICKNAME, username);
+
+  std::string secret(join_secret);
+
+  std::string type = secret.substr(0, secret.find('\n'));
+  size_t offset = type.length() + 1;
+
+  switch (static_cast<SecretType>(std::stol(type)))
+  {
+  default:
+  case SecretType::Empty:
+    return;
+
+  case SecretType::IPAddress:
+  {
+    Config::SetBaseOrCurrent(Config::NETPLAY_TRAVERSAL_CHOICE, "direct");
+
+    std::string host = secret.substr(offset, secret.find_last_of(':') - offset);
+    Config::SetBaseOrCurrent(Config::NETPLAY_HOST_CODE, host);
+
+    offset += host.length();
+    if (secret[offset] == ':')
+      Config::SetBaseOrCurrent(Config::NETPLAY_CONNECT_PORT, std::stoul(secret.substr(offset + 1)));
+  }
+  break;
+
+  case SecretType::RoomID:
+  {
+    Config::SetBaseOrCurrent(Config::NETPLAY_TRAVERSAL_CHOICE, "traversal");
+
+    Config::SetBaseOrCurrent(Config::NETPLAY_HOST_CODE, secret.substr(offset));
+  }
+  break;
+  }
+
+  event_handler->DiscordJoin();
+}
+#endif
+
+Discord::Handler::~Handler() = default;
+
 void Init()
 {
 #ifdef USE_DISCORD_PRESENCE
   if (!Config::Get(Config::MAIN_USE_DISCORD_PRESENCE))
     return;
 
   DiscordEventHandlers handlers = {};
+
+  handlers.ready = HandleDiscordReady;
+  handlers.joinRequest = HandleDiscordJoinRequest;
+  handlers.joinGame = HandleDiscordJoin;
   // The number is the client ID for Dolphin, it's used for images and the appication name
   Discord_Initialize("455712169795780630", &handlers, 1, nullptr);
   UpdateDiscordPresence();
 #endif
 }
 
-void UpdateDiscordPresence()
+void CallPendingCallbacks()
 {
 #ifdef USE_DISCORD_PRESENCE
   if (!Config::Get(Config::MAIN_USE_DISCORD_PRESENCE))
     return;
 
-  const std::string& title = SConfig::GetInstance().GetTitleDescription();
+  Discord_RunCallbacks();
+
+#endif
+}
+
+void InitNetPlayFunctionality(Handler& handler)
+{
+#ifdef USE_DISCORD_PRESENCE
+  event_handler = &handler;
+#endif
+}
+
+void UpdateDiscordPresence(const int party_size, SecretType type, const std::string& secret,
```

Comment:
The "const" in "const int" here does nothing except making it harder to read the signature. Please remove it.